exporter/kafkaexporter/kafka_exporter.go

```diff
@@ -117,122 +105,150 @@ func (e *kafkaMetricsProducer) metricsDataPusher(ctx context.Context, md pmetric
+	return &kafkaExporter[plog.Logs]{
+		cfg:    config,
+		logger: set.Logger,
+		newMessager: func(host component.Host) (kafkaMessager[plog.Logs], error) {
+			marshaler, err := getLogsMarshaler(config.Encoding, host)
+			if err != nil {
+				return nil, err
+			}
+			return &kafkaLogsMessager{
+				config:    config,
+				marshaler: marshaler,
+			}, nil
+		},
 	}
-	return e.producer.Close()
 }
 
-func (e *kafkaLogsProducer) start(ctx context.Context, host component.Host) error {
-	// extensions take precedence over internal encodings
-	if marshaler, errExt := loadEncodingExtension[plog.Marshaler](
-		host,
-		e.cfg.Encoding,
-	); errExt == nil {
-		e.marshaler = &logsEncodingMarshaler{
-			marshaler: *marshaler,
-			encoding:  e.cfg.Encoding,
-		}
-	}
-	if marshaler, errInt := createLogMarshaler(e.cfg); e.marshaler == nil && errInt == nil {
-		e.marshaler = marshaler
-	}
-	if e.marshaler == nil {
-		return errUnrecognizedEncoding
-	}
-	producer, err := newSaramaProducer(ctx, e.cfg)
-	if err != nil {
-		return err
-	}
-	e.producer = producer
-	return nil
+type kafkaLogsMessager struct {
+	config    Config
+	marshaler marshaler.LogsMarshaler
 }
 
-func newSaramaProducer(ctx context.Context, config Config) (sarama.SyncProducer, error) {
-	return kafka.NewSaramaSyncProducer(
-		ctx, config.ClientConfig, config.Producer,
-		config.TimeoutSettings.Timeout,
-	)
+func (e *kafkaLogsMessager) marshalData(ld plog.Logs) ([]marshaler.Message, error) {
+	return e.marshaler.MarshalLogs(ld)
 }
 
-func newMetricsExporter(config Config, set exporter.Settings) *kafkaMetricsProducer {
-	return &kafkaMetricsProducer{
-		cfg:    config,
-		logger: set.Logger,
+func (e *kafkaLogsMessager) getTopic(ctx context.Context, ld plog.Logs) string {
+	return getTopic(ctx, &e.config, ld.ResourceLogs())
+}
+
+func (e *kafkaLogsMessager) partitionData(ld plog.Logs) iter.Seq2[[]byte, plog.Logs] {
+	return func(yield func([]byte, plog.Logs) bool) {
+		if !e.config.PartitionLogsByResourceAttributes {
+			yield(nil, ld)
+			return
+		}
+		for _, resourceLogs := range ld.ResourceLogs().All() {
+			hash := pdatautil.MapHash(resourceLogs.Resource().Attributes())
+			newLogs := plog.NewLogs()
+			resourceLogs.CopyTo(newLogs.ResourceLogs().AppendEmpty())
+			if !yield(hash[:], newLogs) {
+				return
+			}
+		}
 	}
 }
 
-// newTracesExporter creates Kafka exporter.
-func newTracesExporter(config Config, set exporter.Settings) *kafkaTracesProducer {
-	return &kafkaTracesProducer{
+func newMetricsExporter(config Config, set exporter.Settings) *kafkaExporter[pmetric.Metrics] {
+	return &kafkaExporter[pmetric.Metrics]{
 		cfg:    config,
 		logger: set.Logger,
+		newMessager: func(host component.Host) (kafkaMessager[pmetric.Metrics], error) {
+			marshaler, err := getMetricsMarshaler(config.Encoding, host)
+			if err != nil {
+				return nil, err
+			}
+			return &kafkaMetricsMessager{
+				config:    config,
+				marshaler: marshaler,
+			}, nil
+		},
 	}
 }
```

Comment:
Could this be a generic function that is called and set as required?

Reply:
Resolving, let me know if I missed your point